static/app/components/events/viewHierarchy/wireframe.tsx

```diff
@@ -1,62 +1,93 @@
+  const [bounds, setCanvasBounds] = useState<Rect>(Rect.Empty());
+
+  useLayoutEffect(() => {
+    if (!canvases.length) {
+      return undefined;
+    }
+
+    if (canvases.some(c => c === null)) {
+      return undefined;
+    }
+
+    const observer = watchForResize(canvases as HTMLCanvasElement[], entries => {
+      const contentRect =
+        entries[0].contentRect ?? entries[0].target.getBoundingClientRect();
+
+      setCanvasBounds(
+        new Rect(
+          contentRect.x,
+          contentRect.y,
+          contentRect.width * window.devicePixelRatio,
+          contentRect.height * window.devicePixelRatio
+        )
+      );
+    });
+
+    return () => {
+      observer.disconnect();
+    };
+  }, [canvases]);
+
+  return bounds;
+}
 
-const RECT_FILL_ALPHA = 0.005;
-const RECT_OUTLINE_ALPHA = 1;
 const MIN_BORDER_SIZE = 20;
 
-type Rect = Pick<ViewHierarchyWindow, 'x' | 'y' | 'width' | 'height'>;
+interface ViewNode {
+  node: ViewHierarchyWindow;
+  rect: Rect;
+}
 
-export function getCoordinates(hierarchies: ViewHierarchyWindow[]) {
-  return hierarchies.map(hierarchy => {
-    const newHierarchy: Rect[] = [];
+export function getHierarchyDimensions(hierarchies: ViewHierarchyWindow[]): {
+  maxHeight: number;
+  maxWidth: number;
+  nodes: ViewNode[];
+} {
+  const nodes: ViewNode[] = [];
+  const queue: [Rect | null, ViewHierarchyWindow][] = [];
+  [...hierarchies].reverse().forEach(root => queue.push([null, root]));
 
-    const nodesToProcess = [hierarchy];
-    while (nodesToProcess.length) {
-      const node = nodesToProcess.pop()!;
+  let maxWidth = Number.MIN_SAFE_INTEGER;
+  let maxHeight = Number.MIN_SAFE_INTEGER;
 
-      // Fetch coordinates off the current node
-      newHierarchy.push({
-        x: node.x ?? 0,
-        y: node.y ?? 0,
-        width: node.width ?? 0,
-        height: node.height ?? 0,
-      });
+  while (queue.length) {
+    const [parent, child] = queue.pop()!;
 
-      if ((node.children ?? []).length !== 0) {
-        // Update the children's coords relative to the parent
-        const newChildren =
-          node.children?.map(child => ({
-            ...child,
-            x: (child.x ?? 0) + (node.x ?? 0),
-            y: (child.y ?? 0) + (node.y ?? 0),
-          })) ?? [];
+    const node = {
+      node: child,
+      rect: new Rect(
+        (parent?.x ?? 0) + (child.x ?? 0),
+        (parent?.y ?? 0) + (child.y ?? 0),
+        child.width ?? 0,
+        child.height ?? 0
+      ),
+    };
+    nodes.push(node);
 
-        nodesToProcess.push(...newChildren);
-      }
+    if (defined(child.children) && child.children.length) {
+      // Push the children into the queue in reverse order because the
+      // output nodes should have early children before later children
+      // i.e. we need to pop() off early children before ones that come after
+      [...child.children].reverse().forEach(c => {
```

Comment:
when trees get large, doing copies and reverse operations becomes unnecessary overhead and can add up pretty quickly. It's likely not a problem with a few thousand nodes, but you can always avoid this by iterating the children in reverse order